web/packages/teleterm/src/mainProcess/agentDownloader/agentDownloader.ts

```diff
@@ -0,0 +1,168 @@
+ */
+
+import { pipeline } from 'node:stream/promises';
+import { createReadStream } from 'node:fs';
+import { join } from 'node:path';
+import { createUnzip } from 'node:zlib';
+import { execFile } from 'node:child_process';
+import { promisify } from 'node:util';
+
+import { extract } from 'tar-fs';
+
+import { compareSemVers } from 'shared/utils/semVer';
+
+import Logger from 'teleterm/logger';
+
+import { RuntimeSettings } from '../types';
+
+import type { IFileDownloader } from './fileDownloader';
+
+const TELEPORT_CDN_ADDRESS = 'https://cdn.teleport.dev';
+const TELEPORT_RELEASES_ADDRESS = 'https://rlz.teleport.sh/teleport?page=0';
+const logger = new Logger('agentDownloader');
+
+interface AgentBinary {
+  version: string;
+  platform: string;
+  arch: string;
+}
+
+/**
+ * Downloads and unpacks the agent binary, if it has not already been downloaded.
+ *
+ * The agent version to download is taken from settings.appVersion if it is not a dev version (1.0.0-dev).
+ * The settings.appVersion is set to a real version only for packaged apps that went through our CI build pipeline.
+ * In local builds, both for the development version and for packaged apps, settings.appVersion is set to 1.0.0-dev.
+ * In those cases, we fetch the latest available stable version of the agent.
+ * CONNECT_CMC_AGENT_VERSION is available as an escape hatch for cases where we want to fetch a different version.
+ */
+export async function downloadAgent(
+  fileDownloader: IFileDownloader,
+  settings: RuntimeSettings,
+  env: Record<string, any>
+): Promise<void> {
+  const version = await calculateAgentVersion(settings.appVersion, env);
+
+  if (
+    await isCorrectAgentVersionAlreadyDownloaded(
+      settings.agentBinaryPath,
+      version
+    )
+  ) {
+    logger.info(`Agent v${version} is already downloaded. Skipping.`);
+    return;
+  }
+
+  const binaryName = createAgentBinaryName({
+    arch: settings.arch,
+    platform: settings.platform,
+    version,
+  });
+  const url = `${TELEPORT_CDN_ADDRESS}/${binaryName}`;
+  await fileDownloader.run(url, settings.tempDataDir);
+
+  await unpack(join(settings.tempDataDir, binaryName), settings.sessionDataDir);
+
+  logger.info(`Downloaded agent v.${version}.`);
+}
+
+async function calculateAgentVersion(
+  appVersion: string,
+  env: Record<string, any>
+): Promise<string> {
+  if (appVersion !== '1.0.0-dev') {
+    return appVersion;
+  }
+  if (env.CONNECT_CMC_AGENT_VERSION) {
+    return env.CONNECT_CMC_AGENT_VERSION;
+  }
+  return await fetchLatestTeleportRelease();
+}
+
+/**
+ * Takes the first page of teleport releases (30 items) and looks for the highest version.
+ * We don't have a way to simply take the latest tag.
+ */
+async function fetchLatestTeleportRelease(): Promise<string> {
+  const response = await fetch(TELEPORT_RELEASES_ADDRESS);
+  if (!response.ok) {
+    logger.error(response);
+    throw new Error(
+      `Failed to fetch ${TELEPORT_RELEASES_ADDRESS}. Status code: ${response.status}.`
```

Comment:
The `logger` line won't actually log anything, it'll be just `MAIN [agentDownloader] error: {}`.

I was about to suggest that what we throw here is suitable for logs but not really suitable to show to the user. However, it's dev mode only anyway.

I think we can simply remove this `logger.error` line as errors thrown in IPC handlers seem to be logged by Electron out of the box.

Reply:
Ah right, it's not a json object. I removed the line and left the error only.